azurecontainerapps/azurecontainerapps.ts

```diff
@@ -111,15 +100,46 @@ export class azurecontainerapps {
                 }
             }
 
+            // Install the pack CLI if the Oryx builder is being used
+            if (shouldUseBuilder) {
+                await appHelper.installPackCliAsync();
+            }
+
+            // Set the Azure CLI to dynamically install missing extensions
+            util.setAzureCliDynamicInstall();
+
+            // Log in to Azure with the service connection provided
+            const connectedService: string = tl.getInput('connectedServiceNameARM', true);
+            authHelper.loginAzureRM(connectedService);
+
+            const acrUsername: string = tl.getInput('acrUsername', false);
+            const acrPassword: string = tl.getInput('acrPassword', false);
+
+            // Login to ACR if credentials were provided
+            // Note: this step should be skipped if we're ONLY using the YAML configuration file (no image to build/push/pull)
+            if (!shouldOnlyUseYaml && !util.isNullOrEmpty(acrName) && !util.isNullOrEmpty(acrUsername) && !util.isNullOrEmpty(acrPassword)) {
+                console.log(tl.loc('AcrUsernamePasswordLoginMessage'));
+                new ContainerRegistryHelper().loginAcrWithUsernamePassword(acrName, acrUsername, acrPassword);
+            }
+
+            // Login to ACR with access token if complete credentials were not provided
+            // Note: this step should be skipped if we're ONLY using the YAML configuration file (no image to build/push/pull)
+            if (!shouldOnlyUseYaml && !util.isNullOrEmpty(acrName) && (util.isNullOrEmpty(acrUsername) || util.isNullOrEmpty(acrPassword))) {
+                console.log(tl.loc('AcrAccessTokenLoginMessage'));
+                await new ContainerRegistryHelper().loginAcrWithAccessTokenAsync(acrName);
```

Comment:
Nit: probably not needed at this point, but if we end up using the `ContainerRegistryHelper` class in more cases, it might be worth making it `static` so that we don't need to instantiate a new one every time.

Reply:
I originally had a lot of these `Helper` classes as `static`, but when it came to integrating with the test infrastructure in the monorepo, I had to move away from `static` so that I was able to mock the methods in these classes. I have cases where I'm instantiating a single instance of the `Helper` class at the top of the file (such as with `TelemetryHelper`, `AzureAuthenticationHelper` and `ContainerAppHelper`), so it may make sense to instantiate the suite of `Helper` classes right at the beginning to use throughout this `main` method.